File: coconut2d-js/base/CNSprite.js
;
/*!
 *
 *  Coconut-2D JavaScript Library v1.0.0
 *  http://github.com/moky/SpriteForest/coconut2d-js/
 *
 *  Copyright (c) 2014 moKy at slanissue.com
 *  Released under the MIT license
 *  http://en.wikipedia.org/wiki/MIT_License
 *
 */
//
//  CNSprite.js
//  Coconut2D-HTML5
//
//  Created by moKy on Feb. 26, 2014.
//

!function(cn) {
	'use strict';
	var cc = cn.cc;
	
	//
	//  construtor
	//
	function Sprite() {
		cc.Sprite.prototype.ctor.call(this);
	}
	
	//
	//  static functions
	//
	function create() {
		var node = new cn.Sprite();
		if (arguments.length === 0) {
			node.init();
		} else if (typeof arguments[0] === 'string') {
			node.initWithFile(arguments[0]);
		} else if (typeof arguments[0] === 'object') {
			node.initWithDictionary(arguments[0]);
		} else {
			cn.error(arguments);
		}
		return node;
	}
	
	//
	//  instance functions
	//
	function init() {
		if (!cc.Sprite.prototype.init.call(this)) {
			cn.error(this);
			return false;
		}
		
		return true;
	}
	function initWithFile(filename) {
		if (!this.init()) {
			return false;
		}
		this.replaceImageWithFile(filename);
		return true;
	}
	function initWithDictionary(dict) {
		if (!this.init()) {
			return false;
		}
		
		var filename = dict['File'];
		if (filename) {
			this.replaceImageWithFile(filename);
		}
		
		var opacity = dict['Opacity'];
		if (opacity !== undefined) {
			this.setOpacity(opacity);
		}
		
		return this.setAttributes(dict);
	}
	
	function createSprites(filenames) {
		var array = [];
		for (var i = 0; i < filenames.length; ++i) {
			var sprite = create(filenames[i]);
			sprite.getTexture().setAliasTexParameters();
			array.push(sprite);
		}
		return array;
	}
	
	function getImageWithFile(filename) {
		var image = new Image();
		image.src = filename;
		return image;
	}
	
	function getSpriteFrameWithFile(filename) {
		//
		//  try texture cache
		//
		var textureCache = cc.TextureCache.getInstance();
		
		var texture = textureCache.textureForKey(filename);
		if (texture) {
			var size = texture.getContentSize();
			var rect = cc.rect(0, 0, size.width, size.height);
			return cc.SpriteFrame.createWithTexture(texture, rect);
		}
		
		//
		//  try sprite frame cache
		//
		var frameCache = cc.SpriteFrameCache.getInstance();
		
		// try full path
		var frame = frameCache.getSpriteFrame(filename);
		if (frame) {
			return frame;
		}
		
		var pos = filename.lastIndexOf('/');
		if (pos >= 0) {
			// try file name
			frame = frameCache.getSpriteFrame(filename.substring(pos + 1));
			if (frame) {
				return frame;
			}
		}
		
		//
		//  try file
		//
		var fileUtils = cc.FileUtils.getInstance();
		
		// set not pop-up message box when load image failed
		var bNotify = fileUtils.isPopupNotify();
		fileUtils.setPopupNotify(false);
		
		texture = textureCache.addImage(filename);
		if (!texture) {
			var image = getImageWithFile(filename);
			if (image) {
				texture = textureCache.addUIImage(image, filename);
			}
		}
		
		// reset the value of UIImage notify
		fileUtils.setPopupNotify(bNotify);
		
		// got it
		if (texture) {
			var size = texture.getContentSize();
			var rect = cc.rect(0, 0, size.width, size.height);
			return cc.SpriteFrame.createWithTexture(texture, rect);
		}
		
		return null;
	}
	
	function replaceImageWithFile(filename) {
		var spriteFrame = getSpriteFrameWithFile(filename);
		
		if (!spriteFrame) {
			cn.error('failed to load sprite texture: ' + filename);
		} else {
			var texture = spriteFrame.getTexture();
			var rect = spriteFrame.getRect();
			var rotated = spriteFrame.isRotated();
			this.initWithTexture(texture, rect, rotated);
		}
	}
	
	//--------------------------------------------------------------------------
	
	var prop = {
		ctor: Sprite,
		
		init: init,
		initWithFile: initWithFile,
		initWithDictionary: initWithDictionary,
		
		replaceImageWithFile: replaceImageWithFile,
		
		// properties
	};
	
	// extends
	cn.Sprite = cc.Sprite.extend(prop);
	
	// static
	cn.Sprite.create = create;
	cn.Sprite.createSprites = createSprites;
	cn.Sprite.getImageWithFile = getImageWithFile;
	cn.Sprite.getSpriteFrameWithFile = getSpriteFrameWithFile;
	
	// CNNode Extensions
	cn.Node.extensions(cn.Sprite);
	
}(Coconut2D);

File: coconut2d-x/sprites/CNSprite.cpp
//
//  CNSprite.cpp
//  Coconut2D-X
//
//  Created by Moky on 12-11-21.
//  Copyright 2012 Slanissue.com. All rights reserved.
//

#include "CNSprite.h"

#include "../cocoa/CNString.h"
#include "../cocoa/CNFileManager.h"

NS_CN_BEGIN

USING_NS_CC;

bool CNSprite::initWithFile(const char * pszFilename)
{
	if (!init())
	{
		CNLog("error");
		return false;
	}
	replaceImageWithFile(pszFilename);
	return true;
}

bool CNSprite::initWithDictionary(CCDictionary * pDict)
{
	if (!init())
	{
		CNLog("error");
		return false;
	}
	if (!pDict)
	{
		CNLog("invalid parameter: pDict");
		return false;
	}
	
	CCString * pFilename = (CCString *)pDict->objectForKey("File");
	if (pFilename)
	{
		replaceImageWithFile(pFilename->getCString());
	}
	
	CCString * pOpacity = (CCString *)pDict->objectForKey("Opacity");
	if (pOpacity)
	{
		setOpacity(pOpacity->intValue());
	}
	
	return setAttributes(pDict);
}

#pragma mark -

CCArray * CNSprite::createSprites(CCArray * pFilenames)
{
	CCArray * pArray = new CCArray();
	pArray->initWithCapacity(1);
	CNSprite * pSprite = NULL;
	
	CCObject * pObj = NULL;
	CCString * pFilename;
	CCARRAY_FOREACH(pFilenames, pObj)
	{
		pFilename = (CCString *)pObj;
		pSprite = spriteWithFile(pFilename->getCString());
		if (!pSprite)
		{
			CNLog("could not happen!");
			continue;
		}
		pSprite->getTexture()->setAliasTexParameters();
		pArray->addObject(pSprite);
	}
	pArray->autorelease();
	return pArray;
}

/**
 *  change the sprite's image
 */
void CNSprite::replaceImageWithFile(const char * pszFilename)
{
	if (!pszFilename || strlen(pszFilename) == 0)
	{
		return;
	}
	
	CCTextureCache * pTextureCache = CCTextureCache::sharedTextureCache();
	if (!pTextureCache)
	{
		CNLog("could not happen");
		return;
	}
	
	CCTexture2D * pTexture = pTextureCache->textureForKey(pszFilename);
	if (!pTexture)
	{
		// set not pop-up message box when load image failed
		bool bNotify = CCFileUtils::getIsPopupNotify();
		CCFileUtils::setIsPopupNotify(false);
		
		pTexture = pTextureCache->addImage(pszFilename);
		if (!pTexture)
		{
			CCImage * pImage = getImageWithFile(pszFilename);
			if (pImage)
			{
				// create texture from image and put it into the cache
				pTexture = pTextureCache->addUIImage(pImage, pszFilename);
			}
		}
		
		// reset the value of UIImage notify
		CCFileUtils::setIsPopupNotify(bNotify);
	}
	
	bool bRotated = false;
	CCRect tRect = CCRectZero;
	if (pTexture)
	{
		tRect.size = pTexture->getContentSize();
	}
	else
	{
		std::string filename(pszFilename);
		filename = CNString::lastPathComponent(filename);
		CCSpriteFrame * pSpriteFrame = CCSpriteFrameCache::sharedSpriteFrameCache()->spriteFrameByName(filename.c_str());
		if (pSpriteFrame)
		{
			pTexture = pSpriteFrame->getTexture();
			tRect = pSpriteFrame->getRect();
			bRotated = pSpriteFrame->isRotated();
		}
		else
		{
			CNLog("could not get image: %s", pszFilename);
		}
	}
	
	if (pTexture && !CCRect::CCRectEqualToRect(tRect, CCRectZero))
	{
//		setBatchNode(NULL);
//		ccBlendFunc sBlendFunc;
//		sBlendFunc.src = CC_BLEND_SRC;
//		sBlendFunc.dst = CC_BLEND_DST;
//		setBlendFunc(sBlendFunc);
//		
//		setTexture(pTexture);
//		setTextureRect(tRect);
		initWithTexture(pTexture, tRect, bRotated);
	}
}

CCImage * CNSprite::getImageWithFile(const char * pszFilename)
{
	if (!pszFilename || strlen(pszFilename) == 0)
	{
		CNLog("error");
		return NULL;
	}
	if (strstr(pszFilename, "://"))
	{
		// this is a URL
		CNLog("not implement, url: %s", pszFilename);
		return NULL;
	}
	
	std::string filename = CCFileUtils::fullPathFromRelativePath(pszFilename);
	if (!CNFileManager::isExists(pszFilename))
	{
		filename = CNString::lastPathComponent(filename);
		filename = CNFileManager::fullFilePath(filename); // bundle file?
	}
	
	CCImage * pImage = new CCImage();
	if (pImage && pImage->initWithImageFile(filename.c_str()))
	{
		pImage->autorelease();
		return pImage;
	}
	CC_SAFE_DELETE(pImage);
	CNLog("cannot get image from %s", filename.c_str());
	return NULL;
}

NS_CN_END
